tools/cell_census_builder/validate.py

```diff
@@ -281,30 +305,29 @@ def _validate_X_layers_contents_by_dataset(args: Tuple[str, str, Dataset, List[E
                 .soma_joinid.to_numpy()
             )
 
+            raw_sum = 0.0
             raw_nnz = 0
-            if len(soma_joinids) > 0:
-                assert soma.SparseNDArray.exists(exp.ms["RNA"].X["raw"].uri)
-
-                def count_elements(arr: soma.SparseNDArray, join_ids: npt.NDArray[np.int64]) -> int:
-                    return sum(t.non_zero_length for t in arr.read((join_ids, slice(None))).coos())
-
-                raw_nnz = count_elements(exp.ms["RNA"].X["raw"], soma_joinids)
-
-            def count_nonzero(arr: Union[sparse.spmatrix, npt.NDArray[Any]]) -> int:
-                """Return _actual_ non-zero count, NOT the stored value count."""
-                if isinstance(arr, (sparse.spmatrix, sparse.coo_array, sparse.csr_array, sparse.csc_array)):
-                    return np.count_nonzero(arr.data)
-                return np.count_nonzero(arr)
+            fdpm_gene_count = 0
+            if len(cell_soma_joinids) > 0:
+                assert soma.SparseNDArray.exists(exp.ms[MEASUREMENT_RNA_NAME].X["raw"].uri)
+                raw_nnz = _count_elements(exp.ms[MEASUREMENT_RNA_NAME].X["raw"], cell_soma_joinids)
+                raw_sum = _sum_elements(exp.ms[MEASUREMENT_RNA_NAME].X["raw"], cell_soma_joinids)
+                fdpm_gene_count = _count_elements(
+                    exp.ms["RNA"][FEATURE_DATASET_PRESENCE_MATRIX_NAME], [dataset.soma_joinid]
+                )
 
             if ad.raw is None:
-                assert raw_nnz == count_nonzero(
-                    ad.X
-                ), f"{eb.name}:{dataset.dataset_id} 'raw' nnz mismatch {raw_nnz} vs {count_nonzero(ad.X)}"
+                ad_x = ad.X
             else:
-                assert raw_nnz == count_nonzero(
-                    ad.raw.X
-                ), f"{eb.name}:{dataset.dataset_id} 'raw' nnz mismatch {raw_nnz} vs {count_nonzero(ad.raw.X)}"
-
+                ad_x = ad.raw.X
+            h5ad_sum = sum(ad_x.data)  # calculate the actual sum of the X matrix values
```

Comment:
Why test the sum and not test for actual per-element equality?  Equality of sums still allows for a  number of errors (e.g., incorrectly assigned coordinates).  Given that both matrices are loaded and in memory, it is a small step to just compare for actual equality.

In other words, something like:
```
raw_coo = exp.ms[MEASUREMENT_RNA_NAME].X["raw"].read().coos().concat().to_scipy()
assert ad_x.shape == raw_coo.shape
assert (ad_x.tocoo() != raw_coo).nnz == 0
```

It will use some extra memory, but it seems worth it.

A direct equality comparision will provide a super set of both the sum and nnz tests.

Reply:
BTW, if you aren't familiar with the pitfalls of comparing sparse matrices, LMK.  Comparison is an odd beast:
* if the shapes or types differ, it returns True/False
* otherwise it does an element-wise compare (which allocates another matrix of the same size)
So the use of `!=` is a trick - if you expect the compared matrices to be equal, this will allocate a result that is all `False`, which is `0`, which is a very, very sparse matrix (i.e., no memory used).

Reply:
Here's what I tried:
```
raw_coo = exp.ms[MEASUREMENT_RNA_NAME].X["raw"].read((cell_soma_joinids, slice(None))).coos().concat().to_scipy()
assert ad_x.shape == raw_coo.shape
assert (ad_x.tocoo() != raw_coo).nnz == 0
```

The problem is the shape of the ad_x, is different from the raw_coo, and i'm not sure why or how to correct it.

Everything else about the matrixes appears to be correct, nnz, and manually comparing the matrix values looks correct.